src/commons/api/aspa.rs

```diff
@@ -0,0 +1,370 @@
+        }
+        for added in update.added() {
+            if self
+                .providers
+                .iter()
+                .any(|existing| existing.provider() == added.provider())
+            {
+                error.add_duplicate(*added)
+            }
+        }
+
+        if error.is_empty() {
+            Ok(())
+        } else {
+            Err(error)
+        }
+    }
+
+    /// Applies an update. This assumes that the update was verified.
+    pub fn apply_update(&mut self, update: &AspaConfigurationUpdate) {
+        for removed in update.removed() {
+            self.providers.retain(|provider| provider != removed);
+        }
+        for added in update.added() {
+            self.providers.push(*added);
+        }
+        self.providers.sort_by_key(|p| p.provider());
+    }
+}
+
+impl fmt::Display for AspaDefinition {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // example: 65000 => 65001, 65002(v4), 65003(v6)
+        write!(f, "{} => ", self.customer)?;
+        for i in 0..self.providers.len() {
+            if i > 0 {
+                write!(f, ", ")?;
+            }
+            write!(f, "{}", self.providers[i])?;
+        }
+        Ok(())
+    }
+}
+
+impl FromStr for AspaDefinition {
+    type Err = AspaConfigurationFormatError;
+
+    // example: 65000 => 65001, 65002(v4), 65003(v6)
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut parts = s.split("=>");
+
+        let customer = {
+            let customer_str = parts.next().ok_or(AspaConfigurationFormatError::CustomerMissing)?;
+            AspaCustomer::from_str(customer_str.trim())
+                .map_err(|_| AspaConfigurationFormatError::CustomerInvalid(customer_str.trim().to_string()))?
+        };
+
+        let mut providers = {
+            let mut providers = vec![];
+            let providers_str = parts.next().ok_or(AspaConfigurationFormatError::ProvidersMissing)?;
+            let provider_parts = providers_str.split(',');
+            for provider_part in provider_parts {
+                let provider = ProviderAs::from_str(provider_part.trim())
+                    .map_err(|_| AspaConfigurationFormatError::ProviderInvalid(provider_part.trim().to_string()))?;
+                providers.push(provider);
+            }
+            providers
+        };
+
+        // unexpected extra bits are not acceptable
+        if parts.next().is_some() {
+            Err(AspaConfigurationFormatError::ExtraParts)
+        } else {
+            // Ensure that the providers are sorted, there is at least one, and there are no duplicates
+            providers.sort_by_key(|p| p.provider());
+
+            let mut last_seen = providers
+                .first()
+                .ok_or(AspaConfigurationFormatError::ProvidersMissing)?;
+
+            if providers.len() > 1 {
+                for i in 1..providers.len() {
+                    let next = providers.get(i).unwrap(); // safe i goes until .len()
```

Comment:
Doesn't i go until `.len()-1` because you used `a..b` syntax rather than `a..=b` syntax?

Reply:
i should fix the comment - I meant until but not including - which is correct behaviour - the last index number *is* len()-1. Think of an array with 1 entry as an example.. the index for that item is 0. Since I test that the len()>1 I can do all this (start at 1) with fear of  out of bounds.